File: client/src/types/Artifact.ts
export class AverageFeatures {
    danceability: number;
    energy: number;
    speechiness: number;
    valence: number;
    scaledTempo: number;

    constructor(data: AverageFeatures) {
        this.danceability = data.danceability;
        this.energy = data.energy;
        this.speechiness = data.speechiness;
        this.valence = data.valence;
        this.scaledTempo = data.scaledTempo;
    }
}

export class Artifacts {
    songsPlayed: number;
    songsAddedManually: number;
    mostSongsAddedBy: string;
    mostVotesBy: string;
    mostSignificantFeatureOverall: string;
    firstRecommendationVotePercentage: number;
    averageFeatures: AverageFeatures;
    genreStart: string[];
    genreEnd: string[];

    constructor(data: Artifacts) {
        this.songsPlayed = data.songsPlayed;
        this.songsAddedManually = data.songsAddedManually;
        this.mostSongsAddedBy = data.mostSongsAddedBy;
        this.mostVotesBy = data.mostVotesBy;
        this.mostSignificantFeatureOverall = data.mostSignificantFeatureOverall;
        this.firstRecommendationVotePercentage = data.firstRecommendationVotePercentage;
        this.averageFeatures = new AverageFeatures(data.averageFeatures);
        this.genreStart = data.genreStart;
        this.genreEnd = data.genreEnd;
    }
}

File: server/app/models/artifact.py
from typing import Optional

from .camel_model import CamelModel


class AverageFeatures(CamelModel):
    danceability: float
    energy: float
    speechiness: float
    valence: float
    scaled_tempo: float


class Artifact(CamelModel):
    songs_played: int
    songs_added_manually: int
    most_songs_added_by: Optional[list[str]] = None
    most_votes_by: Optional[list[str]] = None
    most_significant_feature_overall: Optional[str] = None
    first_recommendation_vote_percentage: float
    average_features: AverageFeatures
    genre_start: Optional[list[str]] = None
    genre_end: Optional[list[str]] = None
